### ISSUE TITLE
ShowRequestsDialogAsync doesn't loads the friends on W8.1

### ISSUE THREAD

**Issue by CristiW** [Microsoft/winsdkfb#135]
Hello,

I'm trying to invite friends to my app and the Webview used in FacebookDialog.xaml doesn't load all the content.
If I open the request url in a browser, example Microsoft Edge, the friends are correctly loaded.
To authenticate with FB I'm using SessionLoginBehavior::WebAuth to get a long live access token.
![friendrequest](https://cloud.githubusercontent.com/assets/17377076/14081788/074c6c62-f515-11e5-9100-f00b578948cd.png)

**Comment by adiviness** [comment]
`SessionLoginBehavior::WebAuth` does not currently support the feed/requests/send dialogs, it is something that we're actively looking into.

**Comment by adiviness** [comment]
I think we now have a workaround for this problem. You will need to use the branch `webauth_dialogs`. This branch has code that will use a different version of the dialogs when login is performed via the `SessionLoginBehavior::WebAuth` flow. It does come with a couple caveats:

1. The user will be asked to log into Facebook each time a dialog is created unless they check the "remember me" checkbox when they log in (it doesn't need to be the first login, but they will be asked each time until they do check it)
2. There is additional configuration required for the dialogs to close properly and the results of the dialog to be passed back to the sdk. 

Expounding on caveat 2, the WebAuth dialogs expect to be redirected to `ms-app://<your app's SID here>` in order to work properly. If you've set any OAuth redirect URIs in the app's configuration settings in the Facebook developer portal you will need to add an exemption for this ms-app URI otherwise Facebook will block it. Unfortunately, it currently does not seem that you can directly add the ms-app URI to the settings (You won't be able to save the change) so you will instead need to add a url that you control that will redirect to the ms-app URI.

Once you've added a new valid OAuth redirect URI to the app's settings, you need to specify this redirect url to the sdk. This can be done like:
```
FBSession^ sess = FBSession::ActiveSession;
sess->WebAuthDialogRedirectUrl = L"http://your_website.com/your_redirect_page/";
```
You can set this at the same time you're setting up the Windows app ID and the Facebook app ID. 

In order for the correct data to come back from a call to show one of the dialogs (ex. `FBSession::ShowFeedDialogAsync` will return an object that contains the post ID on success) the url that you're redirecting to will need to append the query string of the request to the ms-app URI (so for the Feed dialog you would be redirecting to a url that looks like `ms-app://<your app's SID here>?post_id=<some_number_from_facebook>`). If you don't do this, then the dialog showing function will return an `FBResult` with an error message even if the action was successful. 

Please let us know if this fixes the problem.

**Comment by CristiW** [comment]
Hello,

I didn't tested properly the request functionality because I'm trying to create a redirect page.
However I did find an issue.
In the function DialogUrlBuilder::GetRedirectUriString, if the login is not made explicitly with WebAuth behavior, the redirect is not the one expected.
For example, I did an explicit login with WebAuth behavior and at the next app start, I used Silent behavior to reload the cached access token. In this case, the redirect is not the one webAuth.

Regards,
Cristi

**Comment by adiviness** [comment]
Yes, the new method of displaying the dialogs is only used when `WebAuth` is used as the login method. `Silent` is something that is going to require some more thought for how to integrate it properly. `WebAuth` will not show a dialog if there is a cached access token if you're just looking to avoid the dialog showing in that circumstance. 

To fix that error message, you will need to do some additional configuration in the Facebook settings for the app. In settings => basic, you will need click "Add Platform" and select "Website". The you will need to set both the "App Domains" and the "Site URL" to the website name (ex. `http://example.com`).

**Comment by CristiW** [comment]
Hello adiviness,

I will need more advice from you because the webAuth redirect part is not clear to me.
I created a html page that redirects to the following uri: `ms-app://my app SID` but I'm receiving an error message, as you said in previous post.
How do I append the query string to the ms-app URI? Do I need to modify the redirect URI  for every request? Can you give me an example of how the URI should look, let's say for ShowFeedDialogAsync ?

Regards,
Cristi

**Comment by adiviness** [comment]
Sure. So what's happening it sounds like is that the redirect to the ms-app URI is successful so the dialog closes and the sdk tries to take the query parameters from the response and use them to create an object that wraps up their values (for a feed dialog that object is `FBFeedRequest`). `FBFeedRequest` expects to find a key `post_id` in the response (see the Response Data section of https://developers.facebook.com/docs/sharing/reference/feed-dialog/v2.5). But because that key does not exist you'll get an `FBError` talking about a malformed response. 

In order to fix this you need to take the query string from the request and append it to your response when you perform the redirect. So what you'll be redirecting to should look like `ms-app://<SID here>?post_id=<some long number here>`. How you actually append this data after the ms-app URI depends on the language you're using for the web server. Because the `post_id` will change for each post made with the feed dialog you will need to modify the redirect URI each time.

**Comment by adiviness** [comment]
In the `webauth_dialogs` branch, the`FBSession` will now save the last successful login dialog flow, which should allow the `SessionLoginBehavior::Silent` login parameter to work correctly with the `WebAuthenticationBroker` based dialogs if it is used after `SessionLoginBehavior::WebAuth`.

**Comment by CristiW** [comment]
Hello,

I used html with javascript to create the ms-app redirect url and I hope that the application SID is not sensitive information because it's hardcoded in the html.
Thank you for your help.

Regards,
Cristi

**Comment by MarketMole** [comment]
Please excuse this intrusion. I do not know where else to post/look for this issue. I am no C++ coder, C# only, trying to get winsdkfb into a Universal project. Apparently FBUser.h is missing. Should it be? Should this file by dynamically created perhaps? I figured that the project should compile, and would except that this file is missing. Suggestions?

**Comment by adiviness** [comment]
`FBUser.h` is a file that is generated during compile time. Are you using the nuget package or are you adding a direct reference to the repo? If you are not using the nuget package with a C# project you might see visual studio report some errors falsely because it hasn't finished building the repo yet.

**Comment by MarketMole** [comment]
Not using nuget, direct project inclusion in the solution. 

\facebookwinsdk-master\winsdkfb\winsdkfb\winsdkfb.shared\facebooksession.h(21): fatal error C1083: Cannot open include file: 'FBUser.h': No such file or directory

I get to that and that's it.

**Comment by adiviness** [comment]
Try building the project for x64 first and then back to whatever platform you're originally targeting. 

If that still causes problems, make sure that you cloned the repo and didn't download the zip file instead. Then from a cmd prompt try running `/winsdkfb/winsdkfb/winsdkfb.Shared/codegen.cmd`.

**Comment by MarketMole** [comment]
I did download the zip, but will do a fetch from the repo now... will follow your instructions.

**Comment by MarketMole** [comment]
Ah, there was a previous error that probably was choking the compile:
Could not find TextTransform, please add it to your PATH or have it available in C:\Program Files (x86)\Common Files\Microsoft Shared\TextTemplating\14.0 (VS2015) or C:\Program Files (x86)\Common Files\Microsoft Shared\TextTemplating\12.0 (VS2013)

**Comment by adiviness** [comment]
You will need to have TextTransform.exe in your path in order for codegen.cmd to run correctly, but that should be taken care of for you if you are compiling from within visual studio.

**Comment by MarketMole** [comment]
Thanks for your prompt attention to my comments. Sorry to have abducted this thread. I believe I can get the sample to run now.

**Comment by adiviness** [comment]
No problem. In the future though, making a new github issue would be preferable :)
